File: demo/nodejs/moonphase.js
/*
    moonphase.js  -  by Don Cross - 2019-05-13

    Example Node.js program for Astronomy Engine:
    https://github.com/cosinekitty/astronomy

    This program calculates the Moon's phase for a given date and time,
    or for the computer's current date and time if none is given.
    It also finds the dates and times of the subsequent 10 quarter phase changes.

    To execute, run the command:
    node moonphase [date]
*/

const Astronomy = require('../../source/js/astronomy.js');      // adjust path as needed for your system

function Pad(s, w) {
    s = s.toFixed(0);
    while (s.length < w) {
        s = '0' + s;
    }
    return s;
}

function FormatDate(date) {
    var year = Pad(date.getUTCFullYear(), 4);
    var month = Pad(1 + date.getUTCMonth(), 2);
    var day = Pad(date.getUTCDate(), 2);
    var hour = Pad(date.getUTCHours(), 2);
    var minute = Pad(date.getUTCMinutes(), 2);
    var svalue = date.getUTCSeconds() + (date.getUTCMilliseconds() / 1000);
    var second = Pad(Math.round(svalue), 2);
    return `${year}-${month}-${day} ${hour}:${minute}:${second} UTC`;
}

function Demo() {
    const date = (process.argv.length === 3) ? new Date(process.argv[2]) : new Date();

    // Calculate the Moon's current phase angle, 
    // which ranges from 0 to 360 degrees.
    //   0 degrees = new moon,
    //  90 degrees = first quarter,
    // 180 degrees = full moon,
    // 270 degrees = third quarter.
    const phase = Astronomy.MoonPhase(date);
    console.log(`${FormatDate(date)} : Moon's phase angle = ${phase.toFixed(6)} degrees.`);        
    console.log('');

    // Now we predict when the next 10 lunar quarter phases will happen.
    console.log('The next 10 lunar quarters are:');
    const QuarterName = ['New Moon', 'First Quarter', 'Full Moon', 'Third Quarter'];
    let mq;
    for (let i=0; i < 10; ++i) {
        if (mq === undefined) {
            // The first time around the for loop, we search forward
            // from the current date and time to find the next quarter
            // phase, whatever it might be.
            mq = Astronomy.SearchMoonQuarter(date);
        } else {
            // Use the previous moon quarter information to find the next quarter phase event.
            mq = Astronomy.NextMoonQuarter(mq);
        }
        console.log(`${FormatDate(mq.time.date)} : ${QuarterName[mq.quarter]}`);
    }
}

Demo();
process.exit(0);

File: demo/c/moonphase.c
/*
    moonphase.c  -  by Don Cross - 2019-05-25

    Example C program for Astronomy Engine:
    https://github.com/cosinekitty/astronomy

    This program calculates the Moon's phase for a given date and time,
    or the computer's current date and time if none is given.
    It also finds the dates and times of the subsequent 10 quarter phase changes.
*/

#include <stdio.h>
#include "astro_demo_common.h"

static const char *QuarterName(int quarter)
{
    switch (quarter)
    {
    case 0:   return "New Moon";
    case 1:   return "First Quarter";
    case 2:   return "Full Moon";
    case 3:   return "Third Quarter";
    default:  return "INVALID QUARTER";
    }
}

int main(int argc, const char *argv[])
{
    astro_time_t time;
    astro_angle_result_t phase;
    astro_moon_quarter_t mq;
    astro_illum_t illum;
    int i;

    switch (argc)
    {
    case 1:
        time = Astronomy_CurrentTime();
        break;

    case 2:
        if (ParseTime(argv[1], &time))
            return 1;
        break;

    default:
        fprintf(stderr, "USAGE: moonphase [date]\n");
        return 1;
    }

    /*
        Calculate the Moon's ecliptic phase angle,
        which ranges from 0 to 360 degrees.

          0 = new moon,
         90 = first quarter,
        180 = full moon,
        270 = third quarter.
    */
    phase = Astronomy_MoonPhase(time);
    if (phase.status != ASTRO_SUCCESS)
    {
        printf("Astronomy_MoonPhase error %d\n", phase.status);
        return 1;
    }

    PrintTime(time);
    printf(" : Moon's ecliptic phase angle = %0.3lf degrees.\n", phase.angle);

    /*
        Calculate the percentage of the Moon's disc that is illuminated
        from the Earth's point of view.
    */
    illum = Astronomy_Illumination(BODY_MOON, time);
    if (illum.status != ASTRO_SUCCESS)
    {
        printf("Astronomy_Illumination error %d\n", illum.status);
        return 1;
    }

    PrintTime(time);
    printf(" : Moon's illuminated fraction = %0.2lf%%.", 100.0 * illum.phase_fraction);

    /* Find the next 10 lunar quarter phases. */
    printf("\nThe next 10 lunar quarters are:\n");
    for (i=0; i < 10; ++i)
    {
        if (i == 0)
            mq = Astronomy_SearchMoonQuarter(time);
        else
            mq = Astronomy_NextMoonQuarter(mq);

        if (mq.status != ASTRO_SUCCESS)
        {
            printf("Error %d trying to find moon quarter.\n", mq.status);
            return 1;
        }

        PrintTime(mq.time);
        printf(" : %s\n", QuarterName(mq.quarter));
    }

    return 0;
}
